hydra-plutus/src/Hydra/Contract/Head.hs

```diff
@@ -258,8 +256,13 @@ checkClose ctx parties initialUtxoHash snapshotNumber closedUtxoHash sig cperiod
   outValue =
     maybe mempty (txOutValue . txInInfoResolved) $ findOwnInput ctx
 
+  (closedSnapshotNumber, closedUtxoHash) =
+    case fromBuiltinData @DatumType $ getDatum (continuingDatum ctx) of
+      Just Closed{snapshotNumber, utxoHash} -> (snapshotNumber, utxoHash)
+      _ -> traceError "wrong state in output datum"
+
   checkSnapshot
-    | snapshotNumber == 0 =
+    | closedSnapshotNumber == 0 =
       let expectedOutputDatum =
```

Comment:
I remember vaguely @ch1bo mentioned something about not needing to construct the expected datum here but I could be crazy as well.

Reply:
Yeah.. this code is decoding the `continuingDatum` twice.

Once to get `closedSnapshotNumber` and `closedUtxoHash`, then it builds the expected datum value and compares it against the `continuingDatum`.

Simpler would be to just do the checks we care about. That is
* if `closedSnapshotNumber >= 0`, `verifySnapshotSignature parties closedSnapshotNumber closedUtxoHash sig`
* else (== 0 or negative), `closedUtxoHash == initialUtxoHash`

That should also make the code smaller and more efficient.